### ISSUE TITLE
Clarify what is public and what is private

### ISSUE THREAD

**Issue by asmeurer** [sympy/sympy#23037]
Right now there is not a clear distinction between what is public and what is private in SymPy's API surface. This has led to a "worst of both worlds" state where we default assume everything is public. For example, consider some of the active deprecations (see https://github.com/sympy/sympy/pull/22900, and once that is merged, https://docs.sympy.org/dev/explanation/active-deprecations.html). Quite a few deprecations are for things that clearly should not have been considered public anyway, i.e., we should have been free to make the change without having to worry about a backwards compatibility break. Some examples of this were removing the `_mat` and `_smat` attributes from `Matrix` and `SparseMatrix`, and several deprecations involving moving things to different files (like moving some traversal functions from `sympy.utilities` to `sympy.core`). 

On the other hand, there is the opposite where someone may create a helper function and have had it in their head that it was supposed to be internal only, but then later refactor it only to find that  users found it and were using it.

I expect this issue will be quite challenging to completely figure out. Here are some reasons why this won't be easy

- The only thing that we can say for now is that anything in the top-level namespace (in `sympy/__init__.py`) definitely *is* public. However, this only works one way. There are a lot of things that aren't in the top-level namespace but should still be public. Some things are just not included but probably should be (e.g., any exception that a public SymPy function can raise really ought to be exported). But in other instances entire submodules are explicitly separate from the top-level, like `sympy.stats`, `sympy.codegen`, and so on. There are also a lot of classes that aren't exported but are designed to be subclassed and are therefore public, like the printers. 
- Currently our documentation lists both public and private functions (see https://github.com/sympy/sympy/issues/22105). This means that we can't really say that something is public if it is included in the docs site, because we intentionally document internal functions. We do want to fully split this out so that the main reference part of the docs is only public functions, but this is somewhat of a chicken/egg problem. We can't split out the docs until we know what is public and what is not. 
- A lot of the SymPy API is in methods. Methods are harder because they aren't split by namespace like functions are. Also a lot of methods intentionally start with an underscore, like `_eval_...` methods. Are these considered public or private? This also means that we can't just solve this by saying something is public iff it is in an `__init__.py` (not that I think that's necessarily the best solution anyway).
- It's completely unclear in what instances submodule names are considered part of the public API. If someone does `from sympy.core.function import Function` and we later decide to move `Function` to a different file, or rename `function.py`, is that an API break? If it were something like `sympy.printing.str.StrPrinter`, it would be clear cut, because `StrPrinter` is public API (it's designed to be subclassed), but isn't in the top-level namespace. But for `Function` it is in the top-level namespace. So can we say that people should never be importing it from the specific submodule? I think we should, but I don't know if we've ever made that explicit to users.
- In what cases is the subclass structure of public classes part of the API? We want people to be able to subclass SymPy classes to create their own objects. But, for instance, is the current layout of the matrix subclasses all part of the public API? What about certain superclasses in functions like `TrigonometricFunction`?

Ideally we would have some global way for users to tell what is public and what isn't. Anything that is public would not change in a breaking way. We should then be free to change anything that isn't without deprecation.

Here are some heuristics I wrote at https://github.com/sympy/sympy/pull/22900 (will be at https://docs.sympy.org/dev/guides/contributing/deprecations.html#when-does-a-change-require-deprecation)

**Public API**

- Function names.
- Keyword argument names.
- Keyword argument default values.
- Submodule names.
- The mathematical conventions used to define a function.

**Private API**
- The precise form of an expression. In general, functions may be changed to return a different but mathematically equivalent form of the same expression. This includes a function returning a value which it was not able to compute previously.
- Positional argument names.
- Functions and methods that are private, i.e., for internal use only. Such things should generally be prefixed with an underscore `_`, although this convention is not currently universally adhered to in the SymPy codebase.
- Anything explicitly marked as "experimental".
- Changes to behavior that were mathematically incorrect previously (in general, bug fixes are not considered breaking changes, because despite the saying, bugs in SymPy are not features).
- Anything that was added before the most recent release. Code that has not yet made it into a release does not need to be deprecated. If you are going to change the API of new code, it is best to do it before a release is made so that no deprecations are necessary for future releases.

But even these already have some exceptions, and we haven't really formally agreed upon any of them.

The most obvious answer here is to use the underscore convention, but it isn't clear to me that that will 100% work in all cases. It also somewhat keeps things in the "public by default" state where unless someone remembers to add an underscore to a helper function, it won't be private. I think I might instead prefer some way to explicitly mark things as public, and anything that isn't explicitly marked isn't public. Of course, users could request something that isn't marked to be made public, and we could then take a closer look at it to see if it makes sense. 

And I think it should go without saying, but whatever we decide here should be made with a deprecation so that people that are using things that are currently ambiguous can have a chance to stop using them if they are changed to be explicitly private.

**Comment by ThePauliPrinciple** [comment]
I like the proposed ideas, here are some random thoughts:

> And I think it should go without saying, but whatever we decide here should be made with a deprecation so that people that are using things that are currently ambiguous can have a chance to stop using them if they are changed to be explicitly private.

Would this then involve decorating every `_` function/method out there? 

If there is such a sudden big backwards compatibility change, wouldn't it make sense to change the major version and also tackle other compatibility issues? (e.g. circular imports, the way `Function` works etc).


> It also somewhat keeps things in the "public by default"

Perhaps the CI can report a list of supposed new public and private methods/functions to increase awareness of the choices.


> Here are some heuristics I wrote at

How about considering functions/methods without a docstring as private too?



> * In what cases is the subclass structure of public classes part of the API? We want people to be able to subclass SymPy classes to create their own objects. But, for instance, is the current layout of the matrix subclasses all part of the public API? What about certain superclasses in functions like `TrigonometricFunction`?

I'd say that any `isinstance` or `issubclass` check should still work, but anything else is fine (e.g. adding more subclasses or having a further subclass division in there)

**Comment by asmeurer** [comment]
> Would this then involve decorating every _ function/method out there?

> If there is such a sudden big backwards compatibility change, wouldn't it make sense to change the major version and also tackle other compatibility issues? (e.g. circular imports, the way Function works etc).

I don't know. It depends on how we decide to actually implement this. Using underscores is one idea. Tagging functions is another. Your suggestion of using some tooling on CI to keep track of things is also another idea. 

And it also depends on what fraction of the current API surface we decide is not actually public. If it's a large change then maybe we should do a more major release. That also depends on whether we can actually make this change in the library wholesale, or, as I rather suspect, if we will need to do this bit by bit as we slowly figure this out.

> How about considering functions/methods without a docstring as private too?

I'm strong -1 to using any sort "lack of documentation" as an indicator of public/private. The reason is that all functions and methods should be documented in some way, whether they are public or only intended for internal use. Documenting private methods helps other developers to understand how to use them. A lot of the code in SymPy is inherently complicated and internal documentation helps people be able to contribute code and helps prevent having situations where there is a module that is only understood by a single person.

And on the flip side, there are a lot of public functions that are underdocumented in SymPy right now. We don't want them to be considered private just because no one ever write a docstring for them.

That's also why in https://github.com/sympy/sympy/issues/22105 we want to have some way to list the documentation for all internal functions, but have it separate from the "public API" reference docs so it is still clear to users.

> I'd say that any isinstance or issubclass check should still work, but anything else is fine (e.g. adding more subclasses or having a further subclass division in there)

The question specifically comes up for people making their own subclasses. Should people be able to subclass the Matrix superclasses to make their own custom Matrix class? Or should we only support subclassing the top-level Matrix? It seems like you would want to allow people to subclass things, but that also severely fixes the internal design and limits our ability to refactor it in the future.

**Comment by oscargus** [comment]
I think that introducing the concept that stuff (classes/methods/functions/...) starting with `_` (including modules) are private and should not be relied on for external software is a way to start. Then, we can slowly start transitioning by deprecating (although the new deprecation procedure may be a bit too complicated if there are many of those, introducing a decorator again can probably help out a bit, then one can simply rename to `_` and make the original method/function a deprecated wrapper).

I think this also means that private stuff is not included in the documentation (but should be documented and can/should include doctests). Ideally, one primarily tests public stuff (so call public stuff to trigger private stuff).

What is imported at the top is another thing. I think it makes sense to not clog the name-space too much, so e.g. `stats` can still need to be imported separately.

**Comment by moorepants** [comment]
> Also a lot of methods intentionally start with an underscore, like _eval_... methods.

If we've made special methods that developers (and users) are expected to overwrite, like `__eval` it should be considered public. We probably need a special list of these.

> It's completely unclear in what instances submodule names are considered part of the public API.

I don't think it is unclear. People use the module layout in their downstream code and import directly from modules that contain the functions, classes, and variables they want to use. I agree with @oscargus that we should consider the non-underscored modules to be public and deprecate them if we want to change the name or move them around. We can start now using `_modulename.py` to make new modules non-public. I also don't think we should try moving modules around unless there is a very noteworthy technical need to do so. Although, if we plan a huge layout change of modules for a version 2.0, we could do this, but it's still going to cause so much downstream pain that I don't think it is worth it.

> Ideally we would have some global way for users to tell what is public and what isn't.

I suggest we use prepended `_` to make things public or private, but we maintain a special list of underscored things that we consider public, e.g. `_eval_*`. (We could easily make an `eval_*` method to make it public so we don't need a special list).

> The precise form of an expression. In general, functions may be changed to return a different but mathematically equivalent form of the same expression. This includes a function returning a value which it was not able to compute previously.

For mathematical outputs I agree with this, but we should make it clear that non-mathematical output aspects should not be expected to change. For example and output `{'a': math_expression, 'b': another_math_expression}` should stay a dict with those keywords.

> Positional argument names.

If you have a function with this signature `def f(a, b, c=None):` People can use this in some interesting ways:
```
f(a=1, b=2, c=3) 
f(c=3, b=2, a=1)
kwargs = {a: 1, b: 2, c: 3}
f(**kwargs)
```
So the safe thing is to not changes any argument names.

> Functions and methods that are private, i.e., for internal use only. Such things should generally be prefixed with an underscore _, although this convention is not currently universally adhered to in the SymPy codebase.

Let's do this and spell it out clearly in the developer documentation in big bold words.

> Anything explicitly marked as "experimental".

What is the "marking" for this? I think it should be to use prepended underscores. People can use underscored things but they have to know it may break without deprecation.

> Anything that was added before the most recent release. Code that has not yet made it into a release does not need to be deprecated. If you are going to change the API of new code, it is best to do it before a release is made so that no deprecations are necessary for future releases.

Yes, we should spell this out in the docs so people are aware to watch out for releases and get things in order.

> The most obvious answer here is to use the underscore convention, but it isn't clear to me that that will 100% work in all cases.

I think it work in all cases. We can make things like `_eval_*` no longer have an underscore and make things that need to be private start with one.

> "public by default"

For 15+ years, this is what it has been and I think it is fine. We have to realize that people will use our API in all kinds of unexpected ways. With 1000s and 1000s of users it is so very important to maintain stability of what has always been assumed public. I watch hundreds of new python users in my classes and all I can do is roll my eyes when their code breaks on updates. I am an advanced user and it is ever so tiring and difficult myself to keep my codes running with the continous updates from the scipy ecosystem. If I compare this to matlab, I have never witnessed an issue like this with a student and my code I wrote 20 years ago still runs in the newest matlab with no changes. In 20 years, I've only had to deal with a few matlab api changes and they were minor. Let's hold ourselves to a high standard for the sake of new users, at the minimum.

> How about considering functions/methods without a docstring as private too?

-1 to this. We should document our private code so that we can understand it.

> Should people be able to subclass the Matrix superclasses to make their own custom Matrix class?

If it doesn't have an underscore, people should be able to subclass and rely on non-underscored methods.

> I think that introducing the concept that stuff (classes/methods/functions/...) starting with _ (including modules) are private and should not be relied on for external software is a way to start.

Agreed.

**Comment by smichr** [comment]
> f(c=3, b=2, a=1)

I did not know this could be done! :-()
```python
>>> def f(a,b=2):
...  print(a,b)
... 
>>> f(b=3,a=1)
1 3
```

**Comment by oscarbenjamin** [comment]
> > Also a lot of methods intentionally start with an underscore, like _eval_... methods.
> 
> If we've made special methods that developers (and users) are expected to overwrite, like `_eval` it should be considered public. We probably need a special list of these.

It gets quite complicated with these. I don't know how many downstream libraries implement these methods but I certainly don't think that the ones within the sympy codebase should be considered public in the sense that you should expect to be able to call them directly.

As an example `expr.is_positive` will indirectly call many methods like `expr._eval_is_positive` but no one should call `expr._eval_is_positive` directly: that's handled implicitly by the assumptions system. I have a PR (#22557) which deletes the `Expr._eval_is_positive` method because I can do that without affecting the public result i.e. the front-end `expr.is_positive` will still return the same results. In that sense any `_eval_*` method is private: you shouldn't be calling it directly and you should not even depend on its existence.

However on the flip-side those are the methods that are used to define a Basic subclass and any downstream library might reasonably have also defined methods like `_eval_evalf()`. We should ensure that those methods defined in downstream libraries continue to have the desired effect when participating in the internals of e.g. `Expr.evalf()`.  Preserving that kind of backwards compatibility is hard e.g. it killed this PR: https://github.com/sympy/sympy/pull/20493#issuecomment-735868850

> > "public by default"
> 
> For 15+ years, this is what it has been and I think it is fine.

It's really not fine. There is always a trade-off between compatibility and stagnation. Many things in sympy are quite badly broken and can't really be fixed in a backwards compatible way. Applying backwards compatibility constraints also to all of the things that should really be private is a serious friction against making improvements in sympy.

Perhaps a better approach to take in this issue is rather than saying "clarify what is public and what is private" (the issue title) we could say "clarify what we want to be public and what we want to be private". Then with a process of deprecation we can move things to underscore names and at that point it will be clear what is public and what is private.

I think if I was starting something like sympy from scratch now then the whole `sympy/` directory would be divided with a `sympy._lib` package containing almost all of the code and everything else just being modules whose sole purpose is to define the publicly importable names in the files from which they should be imported. Other libraries seem to be moving to something like that so maybe sympy could as well.

**Comment by moorepants** [comment]
> It gets quite complicated with these.

Then a "special list" of public methods that begin with an underscore is what is likely needed.

> It's really not fine...quite badly broken...

This is in the eye of the beholder. The linux kernel is many times more complicated a crazy than sympy will ever be, yet they don't break user code and my system doesn't fall apart on every kernel update. The code is what it is and just like a historical building or old art painting, you can restore it and improve it without changing it's overall appearance but it takes patience, time, and great effort. Of course, if we can break backwards compatibility it makes it so much easier to make the changes we want to but at the cost of users.

> Applying backwards compatibility constraints also to all of the things that should really be private is a serious friction against making improvements in sympy.

Yes, it is friction, but it isn't a wall. This is the cost of having a popular library. User's don't really care if, for example, the circular dependency issue is fixed by reorangizing all the modules. Maybe it gains them a bit on import time (which we do have complaints about), but they will care if they have to rewrite all their imports. SymPy does work, it may not work in the optimal way from the eyes of a developer, but I think that's ok. We just trudge along patiently and change what we can, when we can like we've always done.

But I'm not wholly opposed to a huge refactor for a SymPy 2.0. If we could get organized and make a big list of clear explanations of these major breaking changes we want to do, then maybe we could execute a SymPy 2.0. But we should keep in mind all the pain Python 2 -> Python 3 caused everyone. It really put tons of stress and extra work on both developers and users. Also, it's quite hard to organize something like this with volunteer sporadic labor. We'd probably need some funding to pay some people to herd this full time.

> Perhaps a better approach to take in this issue is rather than saying "clarify what is public and what is private" (the issue title) we could say "clarify what we want to be public and what we want to be private". Then with a process of deprecation we can move things to underscore names and at that point it will be clear what is public and what is private.

+1

> I think if I was starting something like sympy from scratch now then the whole sympy/ directory would be divided with a sympy._lib package containing almost all of the code and everything else just being modules whose sole purpose is to define the publicly importable names in the files from which they should be imported. Other libraries seem to be moving to something like that so maybe sympy could as well.

We could start that today, but only new things would go in it for now and slowly we could creep old code into it. After 10 years, then maybe sympy would look like that.

**Comment by sylee957** [comment]
The only critical argument I give about encapsulation, is that the reason you had needed 'private' code, is just a sign that you just want to hide away badly documented, badly abstracted, inefficient, badly reusable code.

I don't clearly get the reason why we should encapsulate `_eval_...` stuff for computer algebra systems.
It is just an attempt trying to shy away the procedure 'how to evaluate'.
And in math, it is also important to show how to solve the problem, as much as the answer, but why not we are doing that in SymPy?
